### ISSUE TITLE
Issue with using special characters on the prompts

### ISSUE THREAD

**Issue by Marino252525** [bentleymi/ta-openai-api#18]
Hi,

Thanks so much for this very useful application!

I have a situation, I have created a prompt that will ask the user to post the source code of a program written in Python, but when I put in the code, the system crashes as it is trying to integrate the code into the actual Simple XML code.  I have enclosed the variable in quotes, used the encapsulation so it does not use the characters as Simple XML prompts, but still I having problems.

Here's an example of what I am trying to accomplish:

| openai prompt="Act as a Programmer with 10 years of experience writing code in Python and reviewing the code for bugs.  I need you to review this code '$code.tok$' and see where the bugs are.  Your output will be the analysis of the code and you will re-write it with your recommendations resolving the bugs found. Make sure you include a detailed analysis of your finding to the user learn for future projects."

So, the variable (token) will contain the actual source code in Python, but the prompt will see it as part of the Simple XML code and won't process it.

Any recommendation on how can I encapsulate the token, so it is treated as the text to be analyzed and does not break the program?

Your assistance is greatly appreciated.

**Comment by bentleymi** [comment]
Can you provide an example snippet of code that breaks the system and causes it to crash?

**Comment by Marino252525** [comment]
Hello, certainly, here it is:  


def start_conversation():
    user_input = user_entry.get()
    user_name = "Me"
    bot_name = "GPT"
    conversation = ""
    prompt = user_name + ": " + user_input + "\n" + bot_name+ ": "
    conversation += prompt

    # fetch response from open AI api
    response = openai.Completion.create(
        engine='text-davinci-003', 
        prompt=conversation,
        temperature=1,     
        max_tokens=500)
    response_str = response["choices"][0]["text"].replace("\n", "")
    response_str = response_str.split(user_name + ": ", 1)[0].split(bot_name + ": ", 1)[0]

===========
the error I am getting is:  command="openai", 'NoneType' object has no attribute 'group'.  

Thanks for the quick reply!

**Comment by bentleymi** [comment]
Is this the code you're sending in the prompt or code from the splunk app?

**Comment by Marino252525** [comment]
That is code I am sending in the prompt... basically, this code would be assigned to a variable or token then from there, it will be inserted into the prompt to be sent to the API for processing.

**Comment by bentleymi** [comment]
Perfect, let me test some things and let you know what I figure out.

first thing that comes to mind is possibly URL encoding it first with TA-webtools' "| urlencode " command.

**Comment by Marino252525** [comment]
Hello 

I tried to use urlencode and it did not work.  I installed the TA-webtools to make it available.  Same error:  command="openai", 'NoneType' object has no attribute 'group'

| openai prompt="Can you analyze this source code and improve it: urlencode("import tkinter as tk
from tkinter import ttk
import openai
import pyttsx3
import speech_recognition as sr
from api_key import API_KEY

openai.api_key = API_KEY

engine = pyttsx3.init()

r = sr.Recognizer()
mic = sr.Microphone(device_index=0)

def start_conversation():
    user_input = user_entry.get()
    user_name = "Me"
    bot_name = "GPT"
    conversation = ""
    prompt = user_name + ": " + user_input + "\n" + bot_name+ ": "
    conversation += prompt)"")

**Comment by bentleymi** [comment]
You would need to eval the prompt, urlencode it, then pass it to the map command| eval p="my prompt"| urlencode p| map [ | openai prompt=$p$]On May 20, 2023, at 12:22 PM, Marino252525 ***@***.***> wrote:﻿
Hello
I tried to use urlencode and it did not work.  I installed the TA-webtools to make it available.  Same error:  command="openai", 'NoneType' object has no attribute 'group'
| openai prompt="Can you analyze this source code and improve it: urlencode("import tkinter as tk
from tkinter import ttk
import openai
import pyttsx3
import speech_recognition as sr
from api_key import API_KEY
openai.api_key = API_KEY
engine = pyttsx3.init()
r = sr.Recognizer()
mic = sr.Microphone(device_index=0)
def start_conversation():
user_input = user_entry.get()
user_name = "Me"
bot_name = "GPT"
conversation = ""
prompt = user_name + ": " + user_input + "\n" + bot_name+ ": "
conversation += prompt)"")

**Comment by Marino252525** [comment]
| makeresults 
| eval p="Can you evaluate this program and tell me if there are any bugs on it? 

import tkinter as tk
from tkinter import ttk
import openai
import pyttsx3
import speech_recognition as sr
from api_key import API_KEY

openai.api_key = API_KEY

engine = pyttsx3.init()

r = sr.Recognizer()
mic = sr.Microphone(device_index=0)

def start_conversation():
    user_input = user_entry.get()
    user_name = "Me"
    bot_name = "GPT"
    conversation = ""
    prompt = user_name + ": " + user_input + "\n" + bot_name+ ": "
    conversation += prompt"

| urlencode p
| map  [| openai prompt=$p$]


Now, I get this error:  Error in 'EvalCommand': The expression is malformed.

**Comment by bentleymi** [comment]
When there are quotes in your eval, you have to escape them.  Once you do that, it works without urlencode once you replace [\n|\r]+ with "\n".

```
| makeresults
| eval p="Can you evaluate this program and tell me if there are any bugs on it? 
import tkinter as tk
from tkinter import ttk
import openai
import pyttsx3
import speech_recognition as sr
from api_key import API_KEY

openai.api_key = API_KEY

engine = pyttsx3.init()

r = sr.Recognizer()
mic = sr.Microphone(device_index=0)

def start_conversation():
user_input = user_entry.get()
user_name = \"Me\"
bot_name = \"GPT\"
conversation = \"\"
prompt = user_name + \": \" + user_input + \"\n\" + bot_name+ \": \"
conversation += prompt"
| eval p=replace(p,"[\n|\r]+","\n")
| map [ | openai prompt=$p$]
``` 

![image](https://github.com/bentleymi/ta-openai-api/assets/4107863/eb783b7b-9511-4f5a-891b-ad119dcba2ff)

**Comment by bentleymi** [comment]
I just published version 2.2.0 which allows for a 'prompt_field' option 

ex:
| makeresults count=1
| eval p="My Prompt"
| openai prompt_field=p

You might be interested in that option.

**Comment by bentleymi** [comment]
I corrected a bug in prompt_field implementation with version 2.2.2 today.
